### PR TITLE
Adapter datasource validation

### PR DESCRIPTION
**Pull request by sonallux**
This is the first PR for #156. It only adds validation for the datasource protocol type and format type fields. This is achieved by adding a `Protocol` and a `Format` enum. The PR also removes the `AdapterFactory` to simplify the architecture.

Validation for the other parameters will be added in follow-up PRs.

### CODE REVIEW

**Comment by sonallux** [comment]
@georg-schwarz @andreas-bauer @mathiaszinnen This is ready for review

**Review by georg-schwarz — DISMISSED** [review]
LGTM - I'll let someone else look over it as well..

File: adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/model/Datasource.java [hunk 1]
```diff
@@ -11,123 +12,121 @@
...
-        triggerParameters = runtimeParameters.parameters;
-      }
-      Map<String, String> replacementParameters = Stream.of(defaultParameters, triggerParameters).flatMap(map -> map.entrySet().stream()).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (v1, v2) -> v2));
-      for (Map.Entry<String, String> parameter : replacementParameters.entrySet()) {
-        url = url.replace("{" + parameter.getKey() + "}", parameter.getValue());
-      }
-      HashMap<String, Object> parameters = new HashMap<>(this.getProtocol().getParameters());
-      parameters.put("location", url);
-      return new DatasourceProtocol(this.getProtocol().getType(), parameters);
+  protected Map<String, Object> fillQueryParameters(RuntimeParameters runtimeParameters) {
```
Comment (andreas-bauer) [inline adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/model/Datasource.java:95]: This method looks very crowded.
Could you please add a blank line between the different `Map<>...` statements.

File: adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/model/Datasource.java [hunk 2]
```diff
@@ -11,123 +12,121 @@
...
+    String url = (String) this.getProtocol().getParameters().get("location");
+    Map<String, String> defaultParameters = new HashMap<>();
+    if (this.getProtocol().getParameters().containsKey("defaultParameters")) {
+      defaultParameters = (Map<String, String>) this.getProtocol().getParameters().get("defaultParameters");
     }
+    Map<String, String> triggerParameters = new HashMap<>();
+    if (runtimeParameters != null && runtimeParameters.parameters != null) {
+      triggerParameters = runtimeParameters.parameters;
+    }
+    Map<String, String> replacementParameters = Stream.of(defaultParameters, triggerParameters).flatMap(map -> map.entrySet().stream()).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (v1, v2) -> v2));
```
Comment (andreas-bauer) [inline adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/model/Datasource.java]: This line is too long. You should always split stream API operation into separate lines.
```
Map<String, String> replacementParameters = Stream.of(defaultParameters, triggerParameters)
   .flatMap(map -> map.entrySet().stream())
   .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (v1, v2) -> v2));
```

File: adapter/src/main/java/org/jvalue/ods/adapterservice/adapter/Protocol.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package org.jvalue.ods.adapterservice.adapter;
+
+import org.jvalue.ods.adapterservice.adapter.importer.HttpImporter;
+import org.jvalue.ods.adapterservice.adapter.importer.Importer;
+import org.springframework.web.client.RestTemplate;
+
+public enum Protocol {
```
Comment (andreas-bauer) [inline adapter/src/main/java/org/jvalue/ods/adapterservice/adapter/Protocol.java:7]: Will it work if someone uses lower case `http` or `Http` in the requesting REST API call?

**Comment by andreas-bauer** [comment]
Can `Protocol` and `Format` be in lower case `http` or `json` in a REST request?

**Comment by sonallux** [comment]
> Can `Protocol` and `Format` be in lower case `http` or `json` in a REST request?

No, but it has not worked with lower case before, so nothing has changed from the endpoint user perspective.

If we want to support lower case `http` or even `Http` then we can do that by using a custom Jackson deserializer. But that should be done in a separate PR.

File: adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/model/Datasource.java (hunk 1 above)
Comment (sonallux) [reply adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/model/Datasource.java:95]: Well, this method was there before, I have just adjusted the return type and changed the indentation from 4 to 2 spaces. But I will try to make this method more beautiful.

Comment (andreas-bauer) [reply adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/model/Datasource.java:95]: Hi, I know you didn't created the method. But when you touch it you can also leave it cleaner than before.
The Boy Scout Rule 🙃

**Comment by andreas-bauer** [comment]
Ok, I was just curious. A custom Jackson deserializer just introducing more complexity. As long nobody complains we can keep it this way.

File: adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/model/Datasource.java (hunk 1 above)
Comment (sonallux) [reply adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/model/Datasource.java:95]: I have already refactored it and it is much cleaner now

Comment (andreas-bauer) [reply adapter/src/main/java/org/jvalue/ods/adapterservice/datasource/model/Datasource.java:95]: Looks better now. Thanks

**Review by andreas-bauer — APPROVED** [review]

File: adapter/src/main/java/org/jvalue/ods/adapterservice/adapter/Protocol.java (hunk 3 above)
Comment (mathiaszinnen) [reply adapter/src/main/java/org/jvalue/ods/adapterservice/adapter/Protocol.java:7]: > > Can `Protocol` and `Format` be in lower case `http` or `json` in a REST request?
> 
> No, but it has not worked with lower case before, so nothing has changed from the endpoint user perspective.
> 
> If we want to support lower case `http` or even `Http` then we can do that by using a custom Jackson deserializer. But that should be done in a separate PR.

Couldn't we just extend the existing enum with the other spellings? I agree, however, that we do not need this feature and should leave it out anyway

**Review by mathiaszinnen — APPROVED** [review]
LGTM, too. Much cleaner than before

File: adapter/src/main/java/org/jvalue/ods/adapterservice/adapter/Protocol.java (hunk 3 above)
Comment (sonallux) [reply adapter/src/main/java/org/jvalue/ods/adapterservice/adapter/Protocol.java:7]: Yes, technically that works, too. But I would consider having enums with different spellings as very bad design.

**Pull request merged**